feat(controller): allow enabling pprof (#2838)
akuity/kargo: docs/docs/40-contributor-guide/5-debugging-kargo.md

---
description: Steps to help you debug Kargo when you run into issues.
sidebar_label: Debugging Kargo
---

# Debugging Kargo

From time to time, you may need to debug Kargo. This document outlines options

> to help you to achieve a better understanding of what Kargo (internally) is

Reviewer: ```suggestion
to help you achieve a better understanding of what Kargo (internally) is
```

doing, especially when you are running into issues which are not immediately
obvious by just reading the code.

As a Kargo user, you may not need to debug Kargo itself, but this document may
serve you when a Kargo maintainer asks you to provide additional information
about an issue you are facing.

## Enabling pprof endpoints

> The Kargo controllers can be configured to expose [`pprof`](https://golang.org/pkg/net/http/pprof/)

Reviewer: ```suggestion
Kargo components can be configured to expose [`pprof`](https://golang.org/pkg/net/http/pprof/)
```

> endpoints. These endpoints can be used to profile the Kargo controllers when

Reviewer: ```suggestion
endpoints. These endpoints can be used to profile the components when
```

they are running, and can be useful to understand what the controllers are
doing and where they are spending time.

> To enable the `pprof` endpoint on a Kargo controller, you can set the

Reviewer: ```suggestion
To enable the `pprof` endpoint on a Kargo component, you can set the
```

> `PPROF_BIND_ADDRESS` environment variable to the address where the controller

Reviewer: ```suggestion
`PPROF_BIND_ADDRESS` environment variable to the address where the component
```

should listen for `pprof` requests. For example, to enable the `pprof` endpoint

> on port `6060`, you can set the `PPROF_BIND_ADDRESS` environment variable to

Reviewer: ```suggestion
on port `6060` of the controller, you can set the `PPROF_BIND_ADDRESS` environment variable to
```

`:6060`.

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kargo-controller
spec:
  # ...omitted for brevity
  template:
    spec:
      containers:
      - name: kargo-controller
        env:
        - name: PPROF_BIND_ADDRESS
          value: ":6060"
```

After setting the `PPROF_BIND_ADDRESS` environment variable, the `pprof`
endpoints will be available at `http://<controller-ip>:6060/debug/pprof/`.

### Collecting a profile

To collect a profile, you can port-forward the pprof address to your local
machine and collect the data from an endpoint of choice. For example, to
collect a heap profile, you can run:

```console

> $ kubectl port-forward -n <namespace> deployment/<controller> 6060

Reviewer: ```suggestion
$ kubectl port-forward -n <namespace> deployment/<component> 6060
```
